cdc/redo/meta_manager.go

```diff
@@ -0,0 +1,484 @@
+			if err != nil {
+				return err
+			}
+			metas = append(metas, &meta)
+		}
+		return nil
+	})
+	if err != nil {
+		return errors.WrapError(errors.ErrRedoMetaInitialize,
+			errors.Annotate(err, "read meta file fail"))
+	}
+
+	var checkpointTs, resolvedTs uint64
+	common.ParseMeta(metas, &checkpointTs, &resolvedTs)
+	if checkpointTs == 0 || resolvedTs == 0 {
+		log.Panic("checkpointTs or resolvedTs is 0 when initializing redo meta in owner",
+			zap.Uint64("checkpointTs", checkpointTs),
+			zap.Uint64("resolvedTs", resolvedTs))
+	}
+	m.metaResolvedTs.unflushed = resolvedTs
+	m.metaCheckpointTs.unflushed = checkpointTs
+	if err := m.maybeFlushMeta(ctx); err != nil {
+		return errors.WrapError(errors.ErrRedoMetaInitialize,
+			errors.Annotate(err, "flush meta file fail"))
+	}
+	return util.DeleteFilesInExtStorage(ctx, m.extStorage, toRemoveMetaFiles)
+}
+
+func (m *metaManager) preCleanupExtStorage(ctx context.Context) error {
+	deleteMarker := getDeletedChangefeedMarker(m.changeFeedID)
+	ret, err := m.extStorage.FileExists(ctx, deleteMarker)
+	if err != nil {
+		return errors.WrapError(errors.ErrExternalStorageAPI, err)
+	}
+	if !ret {
+		return nil
+	}
+
+	changefeedMatcher := getChangefeedMatcher(m.changeFeedID)
+	err = util.RemoveFilesIf(ctx, m.extStorage, func(path string) bool {
+		if path == deleteMarker || !strings.Contains(path, changefeedMatcher) {
+			return false
+		}
+		return true
+	}, nil)
+	if err != nil {
+		return err
+	}
+
+	err = m.extStorage.DeleteFile(ctx, deleteMarker)
+	if err != nil && !util.IsNotExistInExtStorage(err) {
+		return errors.WrapError(errors.ErrExternalStorageAPI, err)
+	}
+
+	return nil
+}
+
+// shouldRemoved remove the file which maxCommitTs in file name less than checkPointTs, since
+// all event ts < checkPointTs already sent to sink, the log is not needed any more for recovery
+func (m *metaManager) shouldRemoved(path string, checkPointTs uint64) bool {
+	changefeedMatcher := getChangefeedMatcher(m.changeFeedID)
+	if !strings.Contains(path, changefeedMatcher) {
+		return false
+	}
+	if filepath.Ext(path) != redo.LogEXT {
+		return false
+	}
+
+	commitTs, fileType, err := redo.ParseLogFileName(path)
+	if err != nil {
+		log.Error("parse file name failed", zap.String("path", path), zap.Error(err))
+		return false
+	}
+	if fileType != redo.RedoDDLLogFileType && fileType != redo.RedoRowLogFileType {
+		log.Panic("unknown file type", zap.String("path", path), zap.Any("fileType", fileType))
+	}
+
+	// if commitTs == checkPointTs, the DDL may be executed in the owner,
+	// so we should not delete it.
+	return commitTs < checkPointTs
+}
+
+// deleteAllLogs delete all redo logs and leave a deleted mark.
+func (m *metaManager) deleteAllLogs(ctx context.Context) error {
+	if m.extStorage == nil {
+		return nil
+	}
+	// Write deleted mark before clean any files.
+	deleteMarker := getDeletedChangefeedMarker(m.changeFeedID)
+	err := m.extStorage.WriteFile(ctx, deleteMarker, []byte("D"))
```

Comment:
The `err` here is not checked, is the checking missed?

Reply:
Yes, it missed.  I will fix this.